compiler/src/dotty/tools/dotc/transform/init/Objects.scala

```diff
@@ -0,0 +1,1362 @@
+          // instead of the parameter symbol for the primary constructor.
+          //
+          // abstract class Base(implicit config: Int)
+          // case class A(x: Int)(implicit config: Int) extends Base
+          evalType(sym.termRef, thisV, klass, elideObjectAccess)
+        else
+          readLocal(thisV, sym)
+
+      case tmref: TermRef =>
+        val sym = tmref.symbol
+        if sym.isStaticObject then
+          if elideObjectAccess then
+            ObjectRef(sym.moduleClass.asClass)
+          else
+            accessObject(sym.moduleClass.asClass)
+        else
+          val value = evalType(tmref.prefix, thisV, klass)
+          select(value, tmref.symbol, tmref.prefix)
+
+      case tp @ ThisType(tref) =>
+        val sym = tref.symbol
+        if sym.is(Flags.Package) then
+          Bottom
+        else if sym.isStaticObject && sym != klass then
+          if elideObjectAccess then
+            ObjectRef(sym.moduleClass.asClass)
+          else
+            accessObject(sym.moduleClass.asClass)
+
+        else
+          resolveThis(tref.classSymbol.asClass, thisV, klass)
+
+      case _ =>
+        throw new Exception("unexpected type: " + tp)
+  }
+
+  /** Evaluate arguments of methods and constructors */
+  def evalArgs(args: List[Arg], thisV: Value, klass: ClassSymbol): Contextual[List[ArgInfo]] =
+    val argInfos = new mutable.ArrayBuffer[ArgInfo]
+    args.foreach { arg =>
+      val res =
+        if arg.isByName then
+          Fun(arg.tree, thisV, klass, summon[Env.Data])
+        else
+          eval(arg.tree, thisV, klass)
+
+      val widened =
+        arg.tree.tpe.getAnnotation(defn.InitWidenAnnot) match
+        case Some(annot) =>
+          annot.argument(0).get match
+          case arg @ Literal(c: Constants.Constant) =>
+            val height = c.intValue
+            if height < 0 then
+              report.error("The argument should be positive", arg)
+              res.widen(1)
+            else
+              res.widen(c.intValue)
+          case arg =>
+            report.error("The argument should be a constant integer value", arg)
+            res.widen(1)
+        case _ =>
+          res.widen(1)
+
+      argInfos += TraceValue(widened, trace.add(arg.tree))
+    }
+    argInfos.toList
+
+  /** Initialize part of an abstract object in `klass` of the inheritance chain
+   *
+   * @param tpl       The class body to be evaluated.
+   * @param thisV     The value of the current object to be initialized.
+   * @param klass     The class to which the template belongs.
+   */
+  def init(tpl: Template, thisV: Ref, klass: ClassSymbol): Contextual[Value] = log("init " + klass.show, printer, (_: Value).show) {
+    val paramsMap = tpl.constr.termParamss.flatten.map { vdef =>
+      vdef.name -> Env.valValue(vdef.symbol)
+    }.toMap
+
+    // init param fields
+    klass.paramGetters.foreach { acc =>
+      val value = paramsMap(acc.name.toTermName)
+      if acc.is(Flags.Mutable) then
```

Comment:
I've seen the pattern of this `if` statement before. Perhaps it could be factored out. Or perhaps it's too messy to make that general enough.